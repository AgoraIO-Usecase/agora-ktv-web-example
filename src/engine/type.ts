export enum MusicType {
  /// 快歌
  fast,
  /// 慢歌
  slow,
}

export enum Lang {
  zh = 1,
  en = 2,
  unknown = -1,
}

export interface LyricModel {
  /// 歌曲名称
  name: string;
  /// 歌星名称
  singer: string;
  /// 歌曲类型
  type: MusicType;
  /// 行信息
  lines: LyricLineModel[];
  /// 前奏结束时间  (start开始时间)
  preludeEndPosition: number;
  /// 总时长 (ms) (end 结束时间)
  duration: number;
  /// 是否有pitch值
  hasPitch: boolean;
}

export interface LyricLineModel {
  /// 开始时间 单位为毫秒
  beginTime: number;
  /// 总时长 (ms)
  duration: number;
  /// 行内容
  content: string;
  /// 每行歌词的字信息
  tones: LyricToneModel[];
  // 得分
  score: number;
}

export interface LyricToneModel {
  beginTime: number;
  duration: number;
  word: string;
  pitch: number;
  lang: Lang;
  pronounce: string;
  scores: number[]; // 得分
}

export interface ReactInfo {
  x: number;
  y: number;
  width: number;
}

export enum EnumCombo {
  fair = "fair",
  good = "good",
  excellent = "excellent",
}

export type Combo = {
  type: EnumCombo;
  num: number; // 数量
};

export type Position = {
  x: number;
  y: number;
};

// engine 对外抛的事件
export type EngineEvents = {
  cursor: Position;
  // 每次绘制的数据
  draw: {
    filterReactInfo: ReactInfo[];
    filterColoringReactInfo: ReactInfo[];
  };
  // 展示粒子动画
  cursorTailAnim: boolean;
  // 每秒时间
  timeUpdate: {
    progress: number; // 当前时间 s
    score: number; /// 句子的分数
    lineIndex: number; /// 句子索引
    cumulativeScore: number; /// 累计分数
  };
  // 一句歌词结束
  lineEnd: {
    line: number;
    score: number;
    combo?: Combo;
  };
  hit: {
    word: string; // 命中的字
    stdPitch: number; // 标准音高
    voicePitch: number; // 声音音高 (处理过后)
    originalPitch: number; // 原始音高 (rtc抛出)
  };
};

export enum EnumMessage {
  lyric = 0, // 同步歌词
  pitch = 1, // 同步pitch
  exit = 2, // 退出房间，
  order = 3, // 点歌
}

export type Message = {
  type: EnumMessage;
  [name: string]: any;
};
