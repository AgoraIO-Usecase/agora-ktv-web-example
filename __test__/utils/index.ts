import { apiGetLyric } from "../../src/utils/request.js";

let _dataMap = new Map();

const DEFAULT_URL = "https://accktv.sd-rtn.com/202304072151/734633a36da177df95c3a358677cb3e8/release/lyric/zip_utf8/3/251964.zip";

export const getLyricData = async (url: string = DEFAULT_URL) => {
  if (_dataMap.has(url)) {
    return _dataMap.get(url);
  }
  const res = await apiGetLyric(url);
  _dataMap.set(url, res);
  return res;
};
