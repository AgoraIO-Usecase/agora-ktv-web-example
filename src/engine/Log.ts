export enum EnumLogLevel {
  DEBUG = 0,
  INFO = 1,
  WARN = 2,
  ERROR = 3,
}

export default class Log {
  level: EnumLogLevel;

  constructor() {
    this.level = EnumLogLevel.ERROR;
  }

  // -------- private methods --------
  private _log(...args: any[]) {
    console.log(...args);
  }

  private _err(...args: any[]) {
    let res = "";
    for (let item of args) {
      res += typeof item == "string" ? item : JSON.stringify(item);
    }
    console.error(res);
  }

  //  ----------- public methods -----------
  setLevel(level: EnumLogLevel) {
    if (typeof level != "number" || !(level <= 3 && level >= 0)) {
      throw Error("Invalid level");
    } else {
      this.level = level;
    }
  }

  debug(...args: any[]) {
    this.level <= EnumLogLevel.DEBUG && this._log(`${new Date()}[ENGINE-DEBUG]: `, ...args);
  }

  info(...args: any[]) {
    this.level <= EnumLogLevel.INFO && this._log(`${new Date()}[ENGINE-DEBUG]: `, ...args);
  }

  warn(...args: any[]) {
    this.level <= EnumLogLevel.WARN && this._log(`${new Date()}[ENGINE-DEBUG]: `, ...args);
  }

  error(...args: any[]) {
    this.level <= EnumLogLevel.ERROR && this._err(`${new Date()}[ENGINE-DEBUG]: `, ...args);
  }
}
