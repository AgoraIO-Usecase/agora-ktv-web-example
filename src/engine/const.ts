export const SECOND_PX: number = 80; // 1s等于多少px
export const MILLISECOND_PX: number = SECOND_PX / 1000; // 1ms等于多少px
export const LINE_VIEW_HEIGHT: number = 60; // line view 高度 px
export const LINE_VIEW_WIDTH: number = 375; // line view 宽度 px
export const CURSOR_AREA_WIDTH: number = 100; // 游标左侧区域宽度 px
export const INTERVAL_TIME: number = 20; // 绘制间隔时间
export const FULL_MARK: number = 100; // 满分
export const HIT_SCORE_THRESHOLD: number = 0.7; // 打分偏差
export const COLORING_REACT_TIME = 0.05; // 高亮小块时间 s
