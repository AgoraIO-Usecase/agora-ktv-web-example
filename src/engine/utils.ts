import { LyricLineModel, ReactInfo, LyricToneModel, LyricModel, EnumCombo } from "./type";
import { COLORING_REACT_TIME, SECOND_PX, CURSOR_AREA_WIDTH, HIT_SCORE_THRESHOLD, FULL_MARK } from "./const";

export function getAverage(arr: number[] = []) {
  if (!arr.length) {
    return 0;
  }
  let total = arr.reduce((t, c) => t + c, 0);
  return total / arr.length;
}

export function hasPitch(lines: LyricLineModel[]) {
  for (let line of lines) {
    for (let tone of line.tones) {
      if (tone.pitch) {
        return true;
      }
    }
  }
  return false;
}

export function getTotal(lines: LyricLineModel[]) {
  return lines.length * 100;
}

export function dealPitch(lines: LyricLineModel[]) {
  let min = Number.MAX_VALUE;
  let max = Number.MIN_VALUE;

  for (let line of lines) {
    for (let tone of line.tones) {
      if (tone.pitch < min) {
        min = tone.pitch;
      }
      if (tone.pitch > max) {
        max = tone.pitch;
      }
    }
  }

  return {
    min,
    max,
  };
}

export function cutOffReact(reacts: ReactInfo[], startX: number, endX: number): ReactInfo[] {
  return reacts.map((item) => {
    let start = item.x;
    let end = item.x + item.width;
    let newStart = start;
    let newEnd = end;
    if (start < startX) {
      // 超出左侧
      newStart = startX;
    } else if (end > endX) {
      // 超出右侧
      newEnd = endX;
    }
    return {
      ...item,
      x: newStart - startX, // 重新计算x坐标
      width: newEnd - newStart,
    };
  });
}

export function filterReact(reacts: ReactInfo[], startX: number, endX: number): ReactInfo[] {
  return reacts.filter((item) => {
    const start = item.x;
    const end = start + item.width;
    if (start >= startX && end <= endX) {
      // 在屏幕中
      return true;
    } else if (start <= startX && end >= startX) {
      // 左侧部分超出屏幕
      return true;
    } else if (start <= endX && end >= endX) {
      // 右侧部分超出屏幕
      return true;
    }
    return false;
  });
}

// 判断react是否严格在tone词中
export function isReactInTone(react: ReactInfo, tone: LyricToneModel) {
  const { beginTime, duration } = tone;
  // CURSOR_AREA_WIDTH 油漆区偏移
  const toneStartX = beginTime * SECOND_PX + CURSOR_AREA_WIDTH;
  const toneEndX = toneStartX + duration * SECOND_PX;
  const reactStartX = react.x;
  const reactEndX = react.x + react.width;
  return reactStartX >= toneStartX && reactEndX <= toneEndX;
}

// 间距小于一个react的宽度;
export function isClosedReact(pre: ReactInfo, next: ReactInfo): boolean {
  const distance = COLORING_REACT_TIME * SECOND_PX;
  const preEnd = pre.x + pre.width;
  const nextStart = next.x;
  return Math.abs(nextStart - preEnd) < distance;
}

// 获取击中的react信息 (根据时间)
export function getHitInfo(lyric: LyricModel, currentTime: number): LyricToneModel {
  let tarTone = null;
  for (let line of lyric.lines) {
    for (let tone of line.tones) {
      const begin = tone.beginTime;
      const end = begin + tone.duration;
      if (begin <= currentTime && end >= currentTime) {
        tarTone = tone;
        break;
      }
    }
    if (tarTone) {
      break;
    }
  }
  return tarTone as LyricToneModel;
}

// 获取一行的平均分
export function getLineScore(line: LyricLineModel): number {
  let arr: number[] = [];
  line.tones.forEach((item) => {
    arr.push(getAverage(item.scores));
  });
  return Math.floor(getAverage(arr));
}

// 分数是否命中
export function isHitScore(score: number): boolean {
  return score >= HIT_SCORE_THRESHOLD * FULL_MARK;
}

// 获取一行的文字
export function getLineWords(line: LyricLineModel): string {
  let str = "";
  for (let tone of line.tones) {
    str += tone.word;
  }
  return str;
}

// 获取combo名称
export function getComboName(score: number): EnumCombo | undefined {
  if (score >= 60 && score < 75) {
    return EnumCombo.fair;
  } else if (score >= 75 && score < 90) {
    return EnumCombo.good;
  } else if (score >= 90) {
    return EnumCombo.excellent;
  }
  return undefined;
}
