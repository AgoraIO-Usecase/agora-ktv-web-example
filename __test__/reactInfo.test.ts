import { ReactInfo } from "../src/engine/type";
import { assert, expect, test } from "vitest";
import { filterReact, cutOffReact, isClosedReact } from "../src/engine/utils";
import { CURSOR_AREA_WIDTH } from "../src/engine/const";

const coloringReactInfo1: ReactInfo[] = [
  {
    x: 242.3466666666667,
    y: 33.71559633027523,
    width: 0.5,
  },
  {
    x: 242.8800000000001,
    y: 33.71559633027523,
    width: 0.5,
  },
  {
    x: 243.89333333333298,
    y: 33.71559633027523,
    width: 0.5,
  },
  {
    x: 244.3199999999996,
    y: 33.71559633027523,
    width: 0.37999999999961176,
  },
  {
    x: 245.91999999999985,
    y: 34.678899082568805,
    width: 0.4499999999998394,
  },
  {
    x: 246.34666666666647,
    y: 33.440366972477065,
    width: 0.5,
  },
  {
    x: 248.29333333333295,
    y: 33.30275229357798,
    width: 0.5,
  },
  {
    x: 248.29333333333295,
    y: 33.30275229357798,
    width: 0.5,
  },
  {
    x: 248.90666666666675,
    y: 33.30275229357798,
    width: 0.5,
  },
  {
    x: 250.87999999999965,
    y: 33.85321100917431,
    width: 0.5,
  },
  {
    x: 251.35999999999967,
    y: 33.85321100917431,
    width: 0.5,
  },
  {
    x: 251.8933333333331,
    y: 33.85321100917431,
    width: 0.5,
  },
  {
    x: 252.3199999999997,
    y: 33.85321100917431,
    width: 0.5,
  },
  {
    x: 252.90666666666652,
    y: 33.85321100917431,
    width: 0.5,
  },
  {
    x: 253.33333333333314,
    y: 33.85321100917431,
    width: 0.5,
  },
  {
    x: 259.3599999999998,
    y: 32.20183486238532,
    width: 0.14999999999979252,
  },
  {
    x: 261.8933333333331,
    y: 31.788990825688074,
    width: 0.5,
  },
  {
    x: 262.3466666666667,
    y: 31.788990825688074,
    width: 0.5,
  },
  {
    x: 263.81333333333316,
    y: 31.926605504587158,
    width: 0.5,
  },
  {
    x: 264.3199999999996,
    y: 31.926605504587158,
    width: 0.5,
  },
  {
    x: 264.9066666666664,
    y: 31.926605504587158,
    width: 0.09666666666642953,
  },
  {
    x: 265.78666666666663,
    y: 31.51376146788991,
    width: 0.5,
  },
  {
    x: 266.29333333333307,
    y: 31.51376146788991,
    width: 0.5,
  },
  {
    x: 267.3599999999999,
    y: 31.926605504587158,
    width: 0.5,
  },
  {
    x: 268.29333333333295,
    y: 31.926605504587158,
    width: 0.5,
  },
  {
    x: 268.87999999999977,
    y: 31.926605504587158,
    width: 0.5,
  },
  {
    x: 269.8933333333332,
    y: 31.926605504587158,
    width: 0.5,
  },
  {
    x: 270.34666666666624,
    y: 31.926605504587158,
    width: 0.5,
  },
  {
    x: 279.3599999999998,
    y: 31.651376146788987,
    width: 0.2999999999997982,
  },
  {
    x: 286.29333333333307,
    y: 32.33944954128441,
    width: 0.5,
  },
  {
    x: 292.9066666666665,
    y: 33.02752293577982,
    width: 0.5,
  },
  {
    x: 293.33333333333314,
    y: 33.02752293577982,
    width: 0.5,
  },
  {
    x: 293.893333333333,
    y: 33.02752293577982,
    width: 0.5,
  },
  {
    x: 294.3466666666666,
    y: 33.02752293577982,
    width: 0.5,
  },
  {
    x: 294.88,
    y: 33.02752293577982,
    width: 0.5,
  },
  {
    x: 295.36,
    y: 33.02752293577982,
    width: 0.020000000000024443,
  },
  {
    x: 296.26666666666665,
    y: 32.477064220183486,
    width: 0.5,
  },
  {
    x: 296.26666666666665,
    y: 32.477064220183486,
    width: 0.5,
  },
  {
    x: 312.9066666666665,
    y: 44.31192660550459,
    width: 0.5,
  },
  {
    x: 313.33333333333314,
    y: 44.31192660550459,
    width: 0.5,
  },
  {
    x: 316.8533333333329,
    y: 44.31192660550459,
    width: 0.5,
  },
  {
    x: 342.3466666666667,
    y: 32.752293577981646,
    width: 0.5,
  },
  {
    x: 342.8800000000001,
    y: 32.752293577981646,
    width: 0.5,
  },
  {
    x: 343.893333333333,
    y: 33.30275229357798,
    width: 0.5,
  },
  {
    x: 344.7466666666668,
    y: 33.30275229357798,
    width: 0.5,
  },
  {
    x: 345.33333333333303,
    y: 33.30275229357798,
    width: 0.24333333333302676,
  },
  {
    x: 346.29333333333307,
    y: 33.440366972477065,
    width: 0.5,
  },
  {
    x: 355.36,
    y: 32.20183486238532,
    width: 0.5,
  },
  {
    x: 371.8933333333331,
    y: 31.51376146788991,
    width: 0.5,
  },
  {
    x: 377.91999999999973,
    y: 28.48623853211009,
    width: 0.5,
  },
  {
    x: 378.34666666666635,
    y: 28.48623853211009,
    width: 0.5,
  },
  {
    x: 378.87999999999977,
    y: 28.48623853211009,
    width: 0.15999999999976922,
  },
  {
    x: 380.3199999999998,
    y: 30.825688073394495,
    width: 0.07999999999981355,
  },
  {
    x: 380.90666666666664,
    y: 30.963302752293576,
    width: 0.5,
  },
  {
    x: 381.7599999999999,
    y: 30.963302752293576,
    width: 0.5,
  },
  {
    x: 382.3466666666667,
    y: 32.752293577981646,
    width: 0.5,
  },
  {
    x: 382.8800000000001,
    y: 32.752293577981646,
    width: 0.5,
  },
  {
    x: 403.35999999999956,
    y: 31.788990825688074,
    width: 0.5,
  },
  {
    x: 403.893333333333,
    y: 31.788990825688074,
    width: 0.5,
  },
  {
    x: 404.3199999999996,
    y: 31.788990825688074,
    width: 0.5,
  },
  {
    x: 404.9066666666664,
    y: 31.788990825688074,
    width: 0.2466666666663997,
  },
  {
    x: 411.8933333333331,
    y: 35.22935779816514,
    width: 0.5,
  },
  {
    x: 412.74666666666633,
    y: 35.22935779816514,
    width: 0.5,
  },
  {
    x: 413.91999999999996,
    y: 35.22935779816514,
    width: 0.5,
  },
  {
    x: 414.3466666666666,
    y: 35.22935779816514,
    width: 0.016666666666580454,
  },
  {
    x: 414.88,
    y: 32.752293577981646,
    width: 0.5,
  },
  {
    x: 415.36,
    y: 32.752293577981646,
    width: 0.5,
  },
  {
    x: 415.89333333333343,
    y: 32.752293577981646,
    width: 0.5,
  },
  {
    x: 416.32000000000005,
    y: 32.752293577981646,
    width: 0.5,
  },
  {
    x: 416.9066666666663,
    y: 32.752293577981646,
    width: 0.5,
  },
  {
    x: 417.3333333333329,
    y: 32.752293577981646,
    width: 0.5,
  },
  {
    x: 418.29333333333295,
    y: 32.752293577981646,
    width: 0.5,
  },
  {
    x: 418.29333333333295,
    y: 32.752293577981646,
    width: 0.5,
  },
  {
    x: 419.3599999999998,
    y: 32.752293577981646,
    width: 0.5,
  },
  {
    x: 420.3199999999998,
    y: 32.752293577981646,
    width: 0.5,
  },
  {
    x: 420.90666666666664,
    y: 32.752293577981646,
    width: 0.5,
  },
  {
    x: 421.33333333333326,
    y: 32.752293577981646,
    width: 0.5,
  },
  {
    x: 421.8933333333331,
    y: 32.752293577981646,
    width: 0.5,
  },
  {
    x: 422.3466666666667,
    y: 32.752293577981646,
    width: 0.5,
  },
  {
    x: 422.8800000000001,
    y: 32.752293577981646,
    width: 0.5,
  },
  {
    x: 423.35999999999956,
    y: 32.752293577981646,
    width: 0.5,
  },
  {
    x: 423.893333333333,
    y: 32.752293577981646,
    width: 0.5,
  },
  {
    x: 442.3466666666667,
    y: 31.51376146788991,
    width: 0.5,
  },
  {
    x: 442.8800000000001,
    y: 31.51376146788991,
    width: 0.5,
  },
  {
    x: 443.35999999999956,
    y: 31.51376146788991,
    width: 0.5,
  },
  {
    x: 443.893333333333,
    y: 31.51376146788991,
    width: 0.5,
  },
  {
    x: 444.9066666666664,
    y: 31.51376146788991,
    width: 0.08666666666641731,
  },
  {
    x: 445.33333333333303,
    y: 30.963302752293576,
    width: 0.5,
  },
  {
    x: 451.8666666666667,
    y: 33.1651376146789,
    width: 0.4366666666666674,
  },
  {
    x: 452.3199999999997,
    y: 25.183486238532108,
    width: 0.5,
  },
  {
    x: 454.3466666666666,
    y: 25.183486238532108,
    width: 0.5,
  },
  {
    x: 454.88,
    y: 25.183486238532108,
    width: 0.3899999999999437,
  },
  {
    x: 455.36,
    y: 32.752293577981646,
    width: 0.5,
  },
  {
    x: 465.33333333333303,
    y: 28.89908256880734,
    width: 0.5,
  },
  {
    x: 466.29333333333307,
    y: 28.89908256880734,
    width: 0.5,
  },
  {
    x: 466.8799999999999,
    y: 28.89908256880734,
    width: 0.5,
  },
  {
    x: 468.74666666666656,
    y: 30,
    width: 0.5,
  },
  {
    x: 471.3599999999997,
    y: 30,
    width: 0.08999999999964814,
  },
  {
    x: 471.8666666666667,
    y: 30,
    width: 0.5,
  },
  {
    x: 472.3199999999997,
    y: 30,
    width: 0.5,
  },
  {
    x: 493.33333333333314,
    y: 21.880733944954127,
    width: 0.5,
  },
  {
    x: 493.91999999999996,
    y: 21.880733944954127,
    width: 0.5,
  },
  {
    x: 494.3466666666666,
    y: 21.880733944954127,
    width: 0.5,
  },
  {
    x: 494.88,
    y: 21.880733944954127,
    width: 0.5,
  },
  {
    x: 495.36,
    y: 21.880733944954127,
    width: 0.15999999999998238,
  },
  {
    x: 496.9066666666663,
    y: 26.8348623853211,
    width: 0.5,
  },
  {
    x: 497.3333333333329,
    y: 26.8348623853211,
    width: 0.5,
  },
];

const coloringReactInfo2: ReactInfo[] = [
  {
    x: 498.29333333333295,
    y: 26.8348623853211,
    width: 0.033333333332947745,
  },
  {
    x: 498.29333333333295,
    y: 26.8348623853211,
    width: 0.033333333332947745,
  },
  {
    x: 498.87999999999977,
    y: 27.522935779816514,
    width: 0.5,
  },
  {
    x: 500.74666666666644,
    y: 27.522935779816514,
    width: 0.5,
  },
  {
    x: 500.74666666666644,
    y: 27.522935779816514,
    width: 0.5,
  },
  {
    x: 501.33333333333326,
    y: 27.522935779816514,
    width: 0.02333333333325527,
  },
  {
    x: 501.9200000000001,
    y: 31.238532110091747,
    width: 0.5,
  },
  {
    x: 502.3466666666667,
    y: 31.238532110091747,
    width: 0.5,
  },
  {
    x: 506.8799999999999,
    y: 28.211009174311926,
    width: 0.5,
  },
  {
    x: 507.3599999999999,
    y: 28.211009174311926,
    width: 0.5,
  },
  {
    x: 519.8933333333332,
    y: 22.844036697247706,
    width: 0.5,
  },
  {
    x: 520.3199999999998,
    y: 22.844036697247706,
    width: 0.5,
  },
  {
    x: 521.3333333333333,
    y: 22.844036697247706,
    width: 0.5,
  },
  {
    x: 521.9200000000001,
    y: 22.844036697247706,
    width: 0.2300000000000324,
  },
  {
    x: 522.3466666666667,
    y: 26.69724770642202,
    width: 0.5,
  },
  {
    x: 522.8800000000001,
    y: 26.69724770642202,
    width: 0.5,
  },
  {
    x: 523.3599999999996,
    y: 26.69724770642202,
    width: 0.5,
  },
  {
    x: 524.9066666666664,
    y: 26.69724770642202,
    width: 0.16666666666637298,
  },
  {
    x: 525.333333333333,
    y: 28.34862385321101,
    width: 0.5,
  },
  {
    x: 529.3333333333334,
    y: 31.10091743119266,
    width: 0.5,
  },
  {
    x: 532.9066666666665,
    y: 31.10091743119266,
    width: 0.5,
  },
  {
    x: 533.3333333333331,
    y: 31.10091743119266,
    width: 0.5,
  },
  {
    x: 533.92,
    y: 31.10091743119266,
    width: 0.5,
  },
  {
    x: 536.9066666666663,
    y: 31.926605504587158,
    width: 0.5,
  },
  {
    x: 544.9066666666664,
    y: 29.724770642201836,
    width: 0.5,
  },
  {
    x: 548.7466666666666,
    y: 32.752293577981646,
    width: 0.5,
  },
  {
    x: 549.3333333333334,
    y: 32.752293577981646,
    width: 0.4033333333333644,
  },
  {
    x: 551.3599999999997,
    y: 21.055045871559635,
    width: 0.5,
  },
  {
    x: 551.8933333333331,
    y: 21.055045871559635,
    width: 0.5,
  },
  {
    x: 573.92,
    y: 30.963302752293576,
    width: 0.5,
  },
  {
    x: 574.3466666666666,
    y: 30.963302752293576,
    width: 0.5,
  },
  {
    x: 575.36,
    y: 30.963302752293576,
    width: 0.5,
  },
  {
    x: 586.2933333333331,
    y: 47.064220183486235,
    width: 0.5,
  },
  {
    x: 586.2933333333331,
    y: 47.064220183486235,
    width: 0.5,
  },
  {
    x: 600.7466666666664,
    y: 31.238532110091747,
    width: 0.5,
  },
  {
    x: 603.893333333333,
    y: 36.88073394495413,
    width: 0.5,
  },
  {
    x: 604.7466666666668,
    y: 36.88073394495413,
    width: 0.5,
  },
  {
    x: 604.7466666666668,
    y: 36.88073394495413,
    width: 0.5,
  },
  {
    x: 622.8800000000001,
    y: 33.85321100917431,
    width: 0.5,
  },
  {
    x: 623.3599999999996,
    y: 33.85321100917431,
    width: 0.5,
  },
  {
    x: 623.893333333333,
    y: 33.85321100917431,
    width: 0.5,
  },
  {
    x: 626.2933333333331,
    y: 36.05504587155963,
    width: 0.5,
  },
  {
    x: 626.8799999999999,
    y: 36.05504587155963,
    width: 0.5,
  },
  {
    x: 629.3333333333334,
    y: 36.05504587155963,
    width: 0.5,
  },
  {
    x: 631.8933333333331,
    y: 39.22018348623853,
    width: 0.5,
  },
  {
    x: 632.9066666666665,
    y: 39.22018348623853,
    width: 0.5,
  },
];

const coloringReactInfo3: ReactInfo[] = [
  {
    x: 157.60000000000002,
    y: 35.631768953068594,
    width: 0.5000000000000071,
  },
  {
    x: 158.53333333333336,
    y: 35.631768953068594,
    width: 0.5000000000000071,
  },
];

// 左边超出屏幕
test("select coloring react info from left", () => {
  const startX = 243;
  const endX = startX + CURSOR_AREA_WIDTH;
  let result = filterReact(coloringReactInfo1, startX, endX);
  result = cutOffReact(result, startX, endX);
  expect(result).toMatchSnapshot();
});

// 右边超出屏幕
test("select coloring react info from right", () => {
  const startX = 533;
  const endX = startX + CURSOR_AREA_WIDTH;
  let result = filterReact(coloringReactInfo2, startX, endX);
  result = cutOffReact(result, startX, endX);
  expect(result).toMatchSnapshot();
});

// react 之间的间距大于 react自身宽度
test("react spacing over", () => {
  for (let i = 0; i < coloringReactInfo3.length - 1; i++) {
    const pre = coloringReactInfo3[i];
    const next = coloringReactInfo3[i + 1];
    pre.x = Number(pre.x.toFixed(6));
    pre.width = Number(pre.width.toFixed(6));
    expect(isClosedReact(pre, next)).toBe(true);
  }
});
