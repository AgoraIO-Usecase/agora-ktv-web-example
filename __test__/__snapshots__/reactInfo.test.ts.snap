// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`select coloring react info from left 1`] = `
[
  {
    "width": 0.38000000000010914,
    "x": 0,
    "y": 33.71559633027523,
  },
  {
    "width": 0.5,
    "x": 0.8933333333329756,
    "y": 33.71559633027523,
  },
  {
    "width": 0.37999999999959755,
    "x": 1.3199999999995953,
    "y": 33.71559633027523,
  },
  {
    "width": 0.4499999999998465,
    "x": 2.9199999999998454,
    "y": 34.678899082568805,
  },
  {
    "width": 0.5,
    "x": 3.346666666666465,
    "y": 33.440366972477065,
  },
  {
    "width": 0.5,
    "x": 5.293333333332953,
    "y": 33.30275229357798,
  },
  {
    "width": 0.5,
    "x": 5.293333333332953,
    "y": 33.30275229357798,
  },
  {
    "width": 0.5,
    "x": 5.9066666666667516,
    "y": 33.30275229357798,
  },
  {
    "width": 0.5,
    "x": 7.879999999999654,
    "y": 33.85321100917431,
  },
  {
    "width": 0.5,
    "x": 8.359999999999673,
    "y": 33.85321100917431,
  },
  {
    "width": 0.5,
    "x": 8.89333333333309,
    "y": 33.85321100917431,
  },
  {
    "width": 0.5,
    "x": 9.319999999999709,
    "y": 33.85321100917431,
  },
  {
    "width": 0.5,
    "x": 9.906666666666524,
    "y": 33.85321100917431,
  },
  {
    "width": 0.5,
    "x": 10.333333333333144,
    "y": 33.85321100917431,
  },
  {
    "width": 0.14999999999980673,
    "x": 16.359999999999786,
    "y": 32.20183486238532,
  },
  {
    "width": 0.5,
    "x": 18.89333333333309,
    "y": 31.788990825688074,
  },
  {
    "width": 0.5,
    "x": 19.346666666666692,
    "y": 31.788990825688074,
  },
  {
    "width": 0.5,
    "x": 20.813333333333162,
    "y": 31.926605504587158,
  },
  {
    "width": 0.5,
    "x": 21.319999999999595,
    "y": 31.926605504587158,
  },
  {
    "width": 0.09666666666640822,
    "x": 21.90666666666641,
    "y": 31.926605504587158,
  },
  {
    "width": 0.5,
    "x": 22.786666666666633,
    "y": 31.51376146788991,
  },
  {
    "width": 0.5,
    "x": 23.293333333333067,
    "y": 31.51376146788991,
  },
  {
    "width": 0.5,
    "x": 24.3599999999999,
    "y": 31.926605504587158,
  },
  {
    "width": 0.5,
    "x": 25.293333333332953,
    "y": 31.926605504587158,
  },
  {
    "width": 0.5,
    "x": 25.879999999999768,
    "y": 31.926605504587158,
  },
  {
    "width": 0.5,
    "x": 26.893333333333203,
    "y": 31.926605504587158,
  },
  {
    "width": 0.5,
    "x": 27.346666666666238,
    "y": 31.926605504587158,
  },
  {
    "width": 0.299999999999784,
    "x": 36.359999999999786,
    "y": 31.651376146788987,
  },
  {
    "width": 0.5,
    "x": 43.29333333333307,
    "y": 32.33944954128441,
  },
  {
    "width": 0.5,
    "x": 49.906666666666524,
    "y": 33.02752293577982,
  },
  {
    "width": 0.5,
    "x": 50.333333333333144,
    "y": 33.02752293577982,
  },
  {
    "width": 0.5,
    "x": 50.893333333332976,
    "y": 33.02752293577982,
  },
  {
    "width": 0.5,
    "x": 51.34666666666658,
    "y": 33.02752293577982,
  },
  {
    "width": 0.5,
    "x": 51.879999999999995,
    "y": 33.02752293577982,
  },
  {
    "width": 0.020000000000038654,
    "x": 52.360000000000014,
    "y": 33.02752293577982,
  },
  {
    "width": 0.5,
    "x": 53.26666666666665,
    "y": 32.477064220183486,
  },
  {
    "width": 0.5,
    "x": 53.26666666666665,
    "y": 32.477064220183486,
  },
  {
    "width": 0.5,
    "x": 69.90666666666652,
    "y": 44.31192660550459,
  },
  {
    "width": 0.5,
    "x": 70.33333333333314,
    "y": 44.31192660550459,
  },
  {
    "width": 0.5,
    "x": 73.8533333333329,
    "y": 44.31192660550459,
  },
  {
    "width": 0.5,
    "x": 99.34666666666669,
    "y": 32.752293577981646,
  },
  {
    "width": 0.11999999999989086,
    "x": 99.88000000000011,
    "y": 32.752293577981646,
  },
]
`;

exports[`select coloring react info from right 1`] = `
[
  {
    "width": 0.4066666666665242,
    "x": 0,
    "y": 31.10091743119266,
  },
  {
    "width": 0.5,
    "x": 0.33333333333314386,
    "y": 31.10091743119266,
  },
  {
    "width": 0.5,
    "x": 0.9199999999999591,
    "y": 31.10091743119266,
  },
  {
    "width": 0.5,
    "x": 3.906666666666297,
    "y": 31.926605504587158,
  },
  {
    "width": 0.5,
    "x": 11.90666666666641,
    "y": 29.724770642201836,
  },
  {
    "width": 0.5,
    "x": 15.746666666666556,
    "y": 32.752293577981646,
  },
  {
    "width": 0.40333333333330756,
    "x": 16.33333333333337,
    "y": 32.752293577981646,
  },
  {
    "width": 0.5,
    "x": 18.359999999999673,
    "y": 21.055045871559635,
  },
  {
    "width": 0.5,
    "x": 18.89333333333309,
    "y": 21.055045871559635,
  },
  {
    "width": 0.5,
    "x": 40.91999999999996,
    "y": 30.963302752293576,
  },
  {
    "width": 0.5,
    "x": 41.34666666666658,
    "y": 30.963302752293576,
  },
  {
    "width": 0.5,
    "x": 42.360000000000014,
    "y": 30.963302752293576,
  },
  {
    "width": 0.5,
    "x": 53.29333333333307,
    "y": 47.064220183486235,
  },
  {
    "width": 0.5,
    "x": 53.29333333333307,
    "y": 47.064220183486235,
  },
  {
    "width": 0.5,
    "x": 67.74666666666644,
    "y": 31.238532110091747,
  },
  {
    "width": 0.5,
    "x": 70.89333333333298,
    "y": 36.88073394495413,
  },
  {
    "width": 0.5,
    "x": 71.74666666666678,
    "y": 36.88073394495413,
  },
  {
    "width": 0.5,
    "x": 71.74666666666678,
    "y": 36.88073394495413,
  },
  {
    "width": 0.5,
    "x": 89.88000000000011,
    "y": 33.85321100917431,
  },
  {
    "width": 0.5,
    "x": 90.35999999999956,
    "y": 33.85321100917431,
  },
  {
    "width": 0.5,
    "x": 90.89333333333298,
    "y": 33.85321100917431,
  },
  {
    "width": 0.5,
    "x": 93.29333333333307,
    "y": 36.05504587155963,
  },
  {
    "width": 0.5,
    "x": 93.87999999999988,
    "y": 36.05504587155963,
  },
  {
    "width": 0.5,
    "x": 96.33333333333337,
    "y": 36.05504587155963,
  },
  {
    "width": 0.5,
    "x": 98.89333333333309,
    "y": 39.22018348623853,
  },
  {
    "width": 0.09333333333347582,
    "x": 99.90666666666652,
    "y": 39.22018348623853,
  },
]
`;
